src/hostapi/pulseaudio/pa_linux_pulseaudio_cb.c

```diff
@@ -0,0 +1,718 @@
+            }
+
+            pa_stream_connect_playback( stream->outStream,
+                                        l_ptrPulseAudioHostApi->
+                                        pulseaudioDeviceNames[stream->outDevice],
+                                        &stream->bufferAttr,
+                                        PA_STREAM_INTERPOLATE_TIMING |
+                                        PA_STREAM_ADJUST_LATENCY |
+                                        PA_STREAM_AUTO_TIMING_UPDATE |
+                                        PA_STREAM_NO_REMIX_CHANNELS |
+                                        PA_STREAM_NO_REMAP_CHANNELS |
+                                        PA_STREAM_DONT_MOVE,
+                                        NULL,
+                                        NULL );
+
+            pa_stream_set_underflow_callback( stream->outStream,
+                                              PaPulseAudio_StreamUnderflowCb,
+                                              stream);
+            pa_stream_set_started_callback( stream->outStream,
+                                            PaPulseAudio_StreamStartedCb,
+                                            stream );
+
+            l_strName = NULL;
+        }
+    }
+
+    if( stream->inStream != NULL )
+    {
+        PA_UNLESS(stream->inBuffer =
+                  PaUtil_AllocateMemory(PULSEAUDIO_BUFFER_SIZE),
+                  paInsufficientMemory);
+
+        if( stream->inDevice != paNoDevice)
+        {
+            PA_DEBUG(("Portaudio %s: %d (%s)\n", __FUNCTION__, stream->inDevice,
+                      l_ptrPulseAudioHostApi->pulseaudioDeviceNames[stream->
+                                                                    inDevice]));
+        }
+
+        pa_stream_connect_record( stream->inStream,
+                                  l_ptrPulseAudioHostApi->
+                                  pulseaudioDeviceNames[stream->inDevice],
+                                  &stream->bufferAttr,
+                                  PA_STREAM_INTERPOLATE_TIMING |
+                                  PA_STREAM_ADJUST_LATENCY |
+                                  PA_STREAM_AUTO_TIMING_UPDATE |
+                                  PA_STREAM_NO_REMIX_CHANNELS |
+                                  PA_STREAM_NO_REMAP_CHANNELS |
+                                  PA_STREAM_DONT_MOVE);
+        pa_stream_set_underflow_callback( stream->inStream,
+                                          PaPulseAudio_StreamUnderflowCb,
+                                          stream);
+
+        pa_stream_set_started_callback( stream->inStream,
+                                        PaPulseAudio_StreamStartedCb,
+                                        stream );
+
+    }
+
+    pa_threaded_mainloop_unlock( l_ptrPulseAudioHostApi->mainloop );
+
+    if( stream->outStream != NULL
+        || stream->inStream != NULL )
+    {
+        while( 1)
```

Comment:
This becomes an endless loop if something is wrong. 
I think we should about after a reasonable long time.